receiver/prometheusexecreceiver/receiver_test.go

```diff
@@ -113,3 +120,663 @@ func assertTwoUniqueValuesScraped(t *testing.T, metricsSlice []pdata.Metrics) {
+							},
+						},
+					},
+				},
+			},
+			wantSubprocessConfig: &subprocessmanager.SubprocessConfig{
+				Command: "postgres_exporter",
+				Env: []subprocessmanager.EnvConfig{
+					{
+						Name:  "DATA_SOURCE_NAME",
+						Value: "password:username@(url:port)/dbname",
+					},
+				},
+			},
+		},
+	}
+
+	for _, test := range configTests {
+		t.Run(test.name, func(t *testing.T) {
+			got := getSubprocessConfig(test.config)
+			assert.Equal(t, test.wantSubprocessConfig, got)
+		})
+	}
+}
+
+func TestExtractName(t *testing.T) {
+	customNameTests := []struct {
+		name   string
+		config *Config
+		want   string
+	}{
+		{
+			name: "no custom name",
+			config: &Config{
+				ReceiverSettings: configmodels.ReceiverSettings{
+					TypeVal: "prometheus_exec",
+					NameVal: "prometheus_exec",
+				},
+				ScrapeInterval: 60 * time.Second,
+				Port:           9104,
+				SubprocessConfig: subprocessmanager.SubprocessConfig{
+					Command: "mysqld_exporter",
+					Env:     []subprocessmanager.EnvConfig{},
+				},
+			},
+			want: "prometheus_exec",
+		},
+		{
+			name: "no custom name, only trailing slash",
+			config: &Config{
+				ReceiverSettings: configmodels.ReceiverSettings{
+					TypeVal: "prometheus_exec",
+					NameVal: "prometheus_exec/",
+				},
+				ScrapeInterval: 60 * time.Second,
+				Port:           9104,
+				SubprocessConfig: subprocessmanager.SubprocessConfig{
+					Command: "mysqld_exporter",
+					Env:     []subprocessmanager.EnvConfig{},
+				},
+			},
+			want: "prometheus_exec",
+		},
+		{
+			name: "custom name",
+			config: &Config{
+				ReceiverSettings: configmodels.ReceiverSettings{
+					TypeVal: "prometheus_exec",
+					NameVal: "prometheus_exec/custom",
+				},
+				ScrapeInterval: 60 * time.Second,
+				Port:           9104,
+				SubprocessConfig: subprocessmanager.SubprocessConfig{
+					Command: "mysqld_exporter",
+					Env:     []subprocessmanager.EnvConfig{},
+				},
+			},
+			want: "custom",
+		},
+		{
+			name: "custom name with slashes inside",
+			config: &Config{
+				ReceiverSettings: configmodels.ReceiverSettings{
+					TypeVal: "prometheus_exec",
+					NameVal: "prometheus_exec/custom/name",
+				},
+				ScrapeInterval: 60 * time.Second,
+				Port:           9104,
+				SubprocessConfig: subprocessmanager.SubprocessConfig{
+					Command: "mysqld_exporter",
+					Env:     []subprocessmanager.EnvConfig{},
+				},
+			},
+			want: "custom/name",
+		},
+	}
+
+	for _, test := range customNameTests {
+		t.Run(test.name, func(t *testing.T) {
+			got := extractName(test.config)
+			assert.Equal(t, test.want, got)
+		})
+	}
+}
+
+func TestFillPortPlaceholders(t *testing.T) {
+	fillPortPlaceholdersTests := []struct {
+		name    string
+		wrapper *prometheusExecReceiver
+		newPort int
+		want    *subprocessmanager.SubprocessConfig
+	}{
+		{
+			name: "port is defined by user",
+			wrapper: &prometheusExecReceiver{
+				port: 10500,
+				config: &Config{
+					SubprocessConfig: subprocessmanager.SubprocessConfig{
```

Comment:
Does the subprocess config need to be duplicated here and below? That is quite confusing and makes me think the code is not structured as well as it could be. I think you can remove some duplication here at least (the env values don't need to be specified in both places)?

Reply:
I do need both since the code iterates over the subprocess config ENV variables and replaces them with the placeholder-filled ENV vars from the original config. I can change the code, but in its current form I need at least the ENV structs there to not have an `out of bounds index error` and need the `name` for the test

Reply:
Ok. It's good enough for now but I think this code should definitely be revisited at some point.